apis/python/src/tiledbsoma/collection.py

```diff
@@ -109,6 +110,25 @@ def _create(self, soma_type: str) -> "CollectionBase[CollectionElementType]":
         self._cached_values = {}
         return self
 
+    # lazy tiledb.Group handle for reuse while self is open
+    _open_tiledb_group: Optional[tiledb.Group] = None
+
+    @property
+    def _tiledb_obj(self) -> tiledb.Group:
+        "get the open tiledb.Group handle (opening it if needed)"
+        assert self._open_mode in ("r", "w")
+        if self._open_tiledb_group is None:
+            self._open_tiledb_group = self._close_stack.enter_context(
+                tiledb.Group(
+                    self._uri, mode=self._open_mode, ctx=self.context.tiledb_ctx
+                )
+            )
+        return self._open_tiledb_group
+
+    def close(self) -> None:
+        self._open_tiledb_group = None
+        super().close()  # closes self._open_tiledb_group via self._close_stack
```

Comment:
`_tiledb_obj` and `close` are almost identical between `TileDBArray` and `CollectionBase`. The common logic can be pulled to the base `TileDBObject` by:
1. replacing `_open_tiledb_array` and `_open_tiledb_group` with a `__tiledb_obj` private attribute of TileDBObject (`Union[tiledb.Array, tiledb.Group, None]`)
2. adding a bool class attribute, say `TileDBObject._open_group`, and override it with `TileDBArray._open_group = False` and `CollectionBase._open_group = True`
3. In `TileDBObject._tiledb_obj` use `self._open_group` to determine whether to call `tiledb.open` or `tiledb.Group`

Reply:
I think the subclasses will end up needing to override them anyway to (i) specialize the typing, i.e. `TileDBArray._tiledb_obj -> tiledb.Array` not `Union[tiledb.Array, tiledb.Group] and (ii) to reuse multiple handles as mentioned.

I'm personally averse to the `TileDBObject._open_group` which would imbue the base class with knowledge of its subclasses, not to say that's absolutely forbidden but I think one should usually avoid.

Reply:
It doesn't require knowledge of its subclasses, it just reduces the differences of `_tiledb_object` (or `_sub_open` currently) down to a single boolean.

I guess it comes down to personal preferences, for me DRY code > super strictly-typed code (even more so now that tiledb-py type annotations are not exported; `tiledb.Array == tiledb.Group == Any`)